lib/std/start.zig

```diff
@@ -589,30 +592,95 @@ fn main(c_argc: c_int, c_argv: [*][*:0]c_char, c_envp: [*:null]?[*:0]c_char) cal
-    switch (ReturnType) {
-        void => {
+// Same as std.process.argsAlloc but with null sentinal and c_char
+inline fn cArgvAlloc(arena: std.mem.Allocator) ![:null]?[*:0]c_char {
+    if (@bitSizeOf(c_char) != 8) @compileError("cArgvAlloc is not implemented for non-eight bit c_char");
+
+    var offset = std.ArrayList(usize).init(arena);
+    errdefer offset.deinit();
+    var data = std.ArrayList(u8).init(arena);
+    errdefer data.deinit();
+
+    var args = try process.argsWithAllocator(arena);
+    defer args.deinit();
+    while (args.next()) |arg| {
+        try offset.append(data.items.len);
+        try data.appendSlice(arg[0 .. arg.len + 1]);
+    }
+
+    const addr_owned = try offset.toOwnedSliceSentinel(0);
+    errdefer arena.free(addr_owned);
+    const data_owned = try data.toOwnedSlice();
+    for (addr_owned) |*addr| {
+        addr.* += @intFromPtr(data_owned.ptr);
+    }
+
+    return @ptrCast(addr_owned);
+}
+
+inline fn cEnvpAlloc(arena: std.mem.Allocator) ![:null]?[*:0]c_char {
+    if (@bitSizeOf(c_char) != 8) @compileError("cEnvpAlloc is not implemented for non-eight bit c_char");
+    var envmap = try std.process.getEnvMap(arena);
+    defer envmap.deinit();
+    return @ptrCast(try process.createNullDelimitedEnvMap(arena, &envmap));
+}
+
+pub inline fn callMain(opt_c_argv: ?[:null]?[*:0]c_char, opt_c_envp: ?[*:null]?[*:0]c_char) u8 {
+    const main_fn = @typeInfo(@TypeOf(root.main)).@"fn";
+    const MainReturn = main_fn.return_type.?;
+
+    return switch (MainReturn) {
+        void => blk: {
             root.main();
-            return 0;
+            break :blk 0;
         },
-        noreturn, u8 => {
-            return root.main();
-        },
-        else => {
-            if (@typeInfo(ReturnType) != .error_union) @compileError(bad_main_ret);
+        noreturn, u8 => root.main(),
+        c_int => @intCast(blk: {
+            if (!comptime std.meta.eql(main_fn.calling_convention, .c)) @compileError(bad_c_main);
+
+            if (main_fn.params.len == 0)
+                break :blk root.main();
+
+            var arena_ctx = std.heap.ArenaAllocator.init(std.heap.page_allocator);
+            defer arena_ctx.deinit();
+            const arena = arena_ctx.allocator();
+
+            const c_argv = opt_c_argv orelse (cArgvAlloc(arena) catch |err|
+                std.debug.panic("failed to allocate argv: {s}", .{@errorName(err)}));
+            const c_argc = std.math.cast(c_int, c_argv.len) orelse
+                std.debug.panic("too many arguments passed to main ({d})", .{c_argv.len});
+
+            if (main_fn.params.len == 2)
+                break :blk root.main(c_argc, c_argv.ptr);
+
+            const c_envp = opt_c_envp orelse (cEnvpAlloc(arena) catch |err|
+                std.debug.panic("failed to allocate envp: {s}", .{@errorName(err)})).ptr;
+
+            if (main_fn.params.len == 3)
+                break :blk root.main(c_argc, c_argv.ptr, c_envp);
+
+            @compileError(bad_c_main);
```

Comment:
this looks wrong, none of this was needed before the regression.

Reply:
I agree, taking a quick look at the `0.9.x` branch it didn't have this logic so I wounder how it worked before. I will take a look at this tomorrow.